File: java/PalindromeNumber.java
package leetcode1_10;

/**Determine whether an integer is a palindrome. Do this without extra space.
 * Created by eugene on 16/5/31.
 */
public class PalindromeNumber {

    //先逆转整数,注意overflow
    public boolean isPalindrome(int x) {
        if (x==0) return true;
        if (x<0) return false;
        if (reverse(x)==x) return true;
        return false;
    }
    private static int reverse(int x){
        long res = 0;
        while (x!=0){
            res = res*10 + x%10;
            x = x/10;
            if (res>Integer.MAX_VALUE || res<Integer.MIN_VALUE) return 0;
        }
        return (int)res;
    }

}

File: Python/PalindromeNumber.cpp
using namespace std;

/**Determine whether an integer is a palindrome. Do this without extra space.
 * Coud you solve it without converting the integer to a string?
 * Created by eugene on 16/5/31.
 */
class Solution {
public:
    bool isPalindrome(int x) {
        if(x==0) return true;
        else if(x<0) return false;
        else{
            return x==reverse(x);
        }
    }

    int reverse(int x) {
        long rev = 0;
        while(x!=0){
            rev = rev*10 + x%10;
            x = x/10;
            if(rev>INT_MAX || rev<INT_MIN){
                return 0;
            }
        }
        return static_cast<int>(rev);
    }
};
